qbft/commit.go

```diff
@@ -0,0 +1,201 @@
+package qbft
+
+import (
+	"bytes"
+	"github.com/bloxapp/ssv-spec/types"
+	"github.com/pkg/errors"
+)
+
+// UponCommit returns true if a quorum of commit messages was received.
+func (i *Instance) UponCommit(signedCommit *SignedMessage, commitMsgContainer *MsgContainer) (bool, []byte, *SignedMessage, error) {
+	if i.State.ProposalAcceptedForCurrentRound == nil {
+		return false, nil, nil, errors.New("did not receive proposal for this round")
+	}
+
+	if err := validateCommit(
+		i.config,
+		signedCommit,
+		i.State.Height,
+		i.State.Round,
+		i.State.ProposalAcceptedForCurrentRound,
+		i.State.Share.Committee,
+	); err != nil {
+		return false, nil, nil, errors.Wrap(err, "commit msg invalid")
+	}
+
+	addMsg, err := commitMsgContainer.AddFirstMsgForSignerAndRound(signedCommit)
+	if err != nil {
+		return false, nil, nil, errors.Wrap(err, "could not add commit msg to container")
+	}
+	if !addMsg {
+		return false, nil, nil, nil // UponCommit was already called
+	}
+
+	// calculate commit quorum and act upon it
+	quorum, commitMsgs, err := commitQuorumForRoundValue(i.State, commitMsgContainer, signedCommit.Message.Data, signedCommit.Message.Round)
+	if err != nil {
+		return false, nil, nil, errors.Wrap(err, "could not calculate commit quorum")
+	}
+	if quorum {
+		msgCommitData, err := signedCommit.Message.GetCommitData()
+		if err != nil {
+			return false, nil, nil, errors.Wrap(err, "could not get msg commit data")
+		}
+
+		agg, err := aggregateCommitMsgs(commitMsgs)
```

Comment:
This aggregation of commit messages is a novelty compared to the original specification. It's not clear to me what's the intention behind this and whether it results in a correct protocol.

Reply:
it's just aggregating signatures(in BLS it's very easy to do and verify all aggregated signatures).
Just an easier way to pass a "decided" msg around